libclamav/pe.c

```diff
@@ -2411,125 +2411,146 @@ static inline int hash_impfns(cli_ctx *ctx, void **hashctx, uint32_t *impsz, str
 
 static cl_error_t hash_imptbl(cli_ctx *ctx, unsigned char **digest, uint32_t *impsz, int *genhash, struct cli_exe_info *peinfo)
 {
-    struct pe_image_import_descriptor *image;
+    cl_error_t status = CL_SUCCESS;
+    cl_error_t ret;
+    struct pe_image_import_descriptor image;
+    const struct pe_image_import_descriptor *impdes;
     fmap_t *map = ctx->fmap;
     size_t left, fsize = map->len;
     uint32_t impoff, offset;
-    const char *impdes, *buffer;
+    const char *buffer;
     void *hashctx[CLI_HASH_AVAIL_TYPES];
     enum CLI_HASH_TYPE type;
-    int nimps      = 0;
-    cl_error_t ret = CL_SUCCESS;
+    int nimps = 0;
     unsigned int err;
-    int first = 1;
+    int first          = 1;
+    bool needed_impoff = false;
 
     /* If the PE doesn't have an import table then skip it. This is an
      * uncommon case but can happen. */
     if (peinfo->dirs[1].VirtualAddress == 0 || peinfo->dirs[1].Size == 0) {
         cli_dbgmsg("scan_pe: import table data dir does not exist (skipping .imp scanning)\n");
-        return CL_SUCCESS;
+        goto done;
     }
 
     // TODO Add EC32 wrappers
     impoff = cli_rawaddr(peinfo->dirs[1].VirtualAddress, peinfo->sections, peinfo->nsections, &err, fsize, peinfo->hdr_size);
     if (err || impoff + peinfo->dirs[1].Size > fsize) {
         cli_dbgmsg("scan_pe: invalid rva for import table data\n");
-        return CL_SUCCESS;
+        goto done;
```

Comment:
Does this call for an update to `status`?

Reply:
probably!! nice catch.
Speaking of which, I used to follow a practice of setting the most common failure type for the `status` initialization and then only setting it if a different status was needed... but I've changed my ways and prefer now to be explicit every time I `goto done;` to prevent a bug if someone changes the initialization default.  Please call me out on it if I revert to my old ways.